engine/api/v2_workflow_run_engine.go

```diff
@@ -0,0 +1,384 @@
+		if err := api.Cache.Enqueue(workflow_v2.WorkflowEngineKey, wrEnqueue); err != nil {
+			next()
+			return err
+		}
+		next()
+		return nil
+	}
+	next()
+	defer func() {
+		_ = api.Cache.Unlock(lockKey)
+	}()
+
+	// Load run by id
+	_, next = telemetry.Span(ctx, "workflowRunV2Trigger.LoadRunByID")
+	run, err := workflow_v2.LoadRunByID(ctx, api.mustDB(), wrEnqueue.RunID)
+	if sdk.ErrorIs(err, sdk.ErrNotFound) {
+		next()
+		return nil
+	}
+	if err != nil {
+		next()
+		return sdk.WrapError(err, "unable to load workflow run %s", wrEnqueue.RunID)
+	}
+	next()
+
+	if sdk.StatusIsTerminated(run.Status) && len(wrEnqueue.Jobs) == 0 {
+		log.Debug(ctx, "workflow run already on a final state")
+		return nil
+	}
+
+	u, err := user.LoadByID(ctx, api.mustDB(), wrEnqueue.UserID)
+	if err != nil {
+		return err
+	}
+
+	jobsToQueue, runMsgs, currentJobRunStatus, errRetrieveJobs := retrieveJobToQueue(ctx, api.mustDB(), run, wrEnqueue, u, api.Config.Workflow.JobDefaultRegion)
+	log.Debug(ctx, "workflowRunV2Trigger> jobs to queue: %+v", jobsToQueue)
+
+	tx, errTx := api.mustDB().Begin()
+	if errTx != nil {
+		return sdk.WithStack(err)
+	}
+	defer tx.Rollback() // nolint
+
+	if errRetrieveJobs != nil {
+		for i := range runMsgs {
+			if err := workflow_v2.InsertRunInfo(ctx, tx, &runMsgs[i]); err != nil {
+				return err
+			}
+			run.Status = sdk.StatusFail
+			if err := workflow_v2.UpdateRun(ctx, tx, run); err != nil {
+				return err
+			}
+			if err := tx.Commit(); err != nil {
+				return err
+			}
+		}
+		return errRetrieveJobs
+	}
+
+	// Enqueue JOB
+	for jobID, jobDef := range jobsToQueue {
+		runJob := sdk.V2WorkflowRunJob{
+			WorkflowRunID: run.ID,
+			Status:        sdk.StatusWaiting,
+			JobID:         jobID,
+			Job:           jobDef,
+			UserID:        wrEnqueue.UserID,
+			Username:      u.Username,
+		}
+		if err := workflow_v2.InsertRunJob(ctx, tx, &runJob); err != nil {
+			return err
+		}
+	}
+
+	// Save Run message
+	for i := range runMsgs {
+		if err := workflow_v2.InsertRunInfo(ctx, tx, &runMsgs[i]); err != nil {
+			return err
+		}
+	}
+
+	// End workflow if there is no job to queue,  no running jobs and current status is not terminated
+	if len(jobsToQueue) == 0 && sdk.StatusIsTerminated(currentJobRunStatus) && !sdk.StatusIsTerminated(run.Status) {
+		run.Status = currentJobRunStatus
+
+		if err := workflow_v2.UpdateRun(ctx, tx, run); err != nil {
+			return err
+		}
+	}
+
+	return sdk.WithStack(tx.Commit())
+}
+
+// TODO manage re run
+func retrieveJobToQueue(ctx context.Context, db *gorp.DbMap, run *sdk.V2WorkflowRun, wrEnqueue sdk.V2WorkflowRunEnqueue, u *sdk.AuthentifiedUser, defaultRegion string) (map[string]sdk.V2Job, []sdk.V2WorkflowRunInfo, string, error) {
+	runInfos := make([]sdk.V2WorkflowRunInfo, 0)
+	jobToQueue := make(map[string]sdk.V2Job)
+
+	// Load run_jobs
+	_, next := telemetry.Span(ctx, "retrieveJobToQueue.LoadRunJobsByRunID")
+	runJobs, err := workflow_v2.LoadRunJobsByRunID(ctx, db, run.ID)
```

Comment:
put the span in the function retrieveJobToQueue.LoadRunJobsByRunID

Reply:
done